[4796] Reactify Code Editor

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/4796

aachavarria committed:
Creating CodeEditorDialog.tsx

aachavarria committed:
Merge branch 'feature/4773' of https://github.com/aachavarria/studio-ui into feature/4796

aachavarria committed:
Adding snippets;
Adding Save Actions;
Creating writeContent;

aachavarria committed:
adding mode to CreateFileDialog.tsx;

aachavarria committed:
Fixing issues related with editContentType

aachavarria committed:
creating variables names for groovy files

aachavarria committed:
deleting comments;

aachavarria committed:
reverting reloadDetailedItem after unlock item;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4796

aachavarria committed:
Fixing issues related with re-fetching the content after close the dialog

aachavarria committed:
adding DialogProps to CodeEditorDialogProps
adding saving shortCut
adding ConditionallyUnlockItem

aachavarria committed:
Fixing issues related with updates on showCodeEditorDialog

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialog.tsx:
```diff
@@ -0,0 +1,99 @@
+import { CodeEditorDialogContainer } from './CodeEditorDialogContainer';
+
+interface CodeEditorDialogBaseProps {
+  open?: boolean;
```
should not be optional

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialog.tsx:
```diff
@@ -0,0 +1,99 @@
+        })
+      );
+    } else {
+      dispatch(ConditionallyUnlockItem({ path: props.path, notify: false }));
```
action name should not be capitalised

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialog.tsx:
```diff
@@ -0,0 +1,99 @@
+  };
+
+  return (
+    <Dialog open={open && !minimized} keepMounted={minimized} onClose={onClose} {...rest} fullWidth maxWidth="xl">
```
fullwidth and maxwidth can be overridden too
```suggestion
    <Dialog open={open && !minimized} keepMounted={minimized} onClose={onClose} fullWidth maxWidth="xl" {...rest}>
```

rart commented on ui/app/src/utils/string.ts:
```diff
@@ -67,6 +67,10 @@ export function isJSON(str: string): boolean {
   throw new Error('[isJSON] Not implemented.');
 }
 
+export function isCamelCase(str: string) {
```
This doesn't test for camel case.
```suggestion
export function hasUppercaseChars(str: string) {
```

rart commented on ui/app/src/utils/hooks.ts:
```diff
@@ -174,6 +175,17 @@ export function useItemsByPath(): GlobalState['content']['itemsByPath'] {
+    if (nou(item) && path) {
+      dispatch(completeDetailedItem({ path }));
+    }
+  }, [dispatch, item, path]);
```
We need a mechanism to make sure we don't dispatch a million of these calls due to subsequent renders when the item hasn't loaded yet.

rart commented on ui/app/src/state/epics/content.ts:
```diff
@@ -201,6 +202,32 @@ const content: CrafterCMSEpic[] = [
+  // region ConditionallyUnlockItem
+  (action$, state$) =>
+    action$.pipe(
+      ofType(ConditionallyUnlockItem.type),
```
```suggestion
  // region conditionallyUnlockItem
  (action$, state$) =>
    action$.pipe(
      ofType(conditionallyUnlockItem.type),
```

rart commented on ui/app/src/state/epics/content.ts:
```diff
@@ -201,6 +202,32 @@ const content: CrafterCMSEpic[] = [
+            )
+          );
+        } else {
+          return NEVER;
```
`filter` instead of `NEVER`

rart commented on ui/app/src/state/actions/content.ts:
```diff
@@ -80,3 +80,7 @@ export const pasteItemWithPolicyValidation = /*#__PURE__*/ createAction<{ path:
 
 export const unlockItem = /*#__PURE__*/ createAction<{ path: string; notify?: boolean }>('UNLOCK_ITEM');
+
+export const ConditionallyUnlockItem = /*#__PURE__*/ createAction<{ path: string; notify?: boolean }>(
```
```suggestion
export const conditionallyUnlockItem = /*#__PURE__*/ createAction<{ path: string; notify?: boolean }>(
```

rart replied:
Add a bit of a comment explaining what the action does — explain it's conditionality

rart commented on ui/app/src/services/content.ts:
```diff
@@ -131,6 +131,38 @@ export function fetchContentInstance(
+  user: string,
+  fileName: string,
+  content: string,
+  unlock: boolean = true
```
Please check if `user` is optional for the service. Let's try to modify the abstraction to be like this:
```suggestion
  site: string,
  path: string,
  content: string,
  options?: { user?: string, unlock?: boolean }
```

rart commented on ui/app/src/components/CodeEditorDialog/utils.ts:
```diff
@@ -0,0 +1,131 @@
+  'get-spring-bean': {
+    label: 'Get Spring Bean',
+    value: 'applicationContext.get("SPRING_BEAN_NAME")'
+  }
```
Please move snippets to ui.xml

@sumerjabri @russdanner fyi

aachavarria committed:
renaming ConditionallyUnlockItem to conditionallyUnlockItem;
renaming isCamelCase to hasUppercaseChars;
using filter on conditionallyUnlockItem epic

aachavarria committed:
adding comment to conditionallyUnlockItem action

aachavarria committed:
creating getFileNameFromPath util
updating writeContent service

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+      <Menu anchorEl={anchorEl} keepMounted open={Boolean(anchorEl)} onClose={closeSnippets}>
+        {contentModelSnippets && (
+          <ListSubheader disableSticky={true}>
+            <FormattedMessage id="codeEditor" defaultMessage="Content model" />
```
bad message id

rart commented on ui/app/src/services/content.ts:
```diff
@@ -131,6 +131,31 @@ export function fetchContentInstance(
+      if (ajaxResponse.response.result.error) {
+        // eslint-disable-next-line no-throw-literal
+        throw {
+          ...ajaxResponse,
```
fake out the `500` status

rart commented on ui/app/src/state/actions/content.ts:
```diff
@@ -80,3 +80,8 @@ export const pasteItemWithPolicyValidation = /*#__PURE__*/ createAction<{ path:
 
 export const unlockItem = /*#__PURE__*/ createAction<{ path: string; notify?: boolean }>('UNLOCK_ITEM');
+
+// To unlock a item the item.lockOwner should be empty or should be equals to the current user, this action check if can be unlocked
```
Why would you issue an unlock call for an item that's already unlocked?
```suggestion
// This action's semantic is to unlock the item only if the lock owner is the current user.
```

rart commented on ui/app/src/state/epics/content.ts:
```diff
@@ -201,6 +202,31 @@ const content: CrafterCMSEpic[] = [
+      filter(
+        ([{ payload }, state]) =>
+          !state.content.itemsByPath[payload.path].lockOwner ||
+          state.content.itemsByPath[payload.path].lockOwner === state.user.username
```
Why would you issue an unlock call for an item that's already unlocked?
```suggestion
        ([{ payload }, state]) =>
          state.content.itemsByPath[payload.path].lockOwner === state.user.username
```

rart commented on ui/app/src/state/epics/content.ts:
```diff
@@ -201,6 +202,31 @@ const content: CrafterCMSEpic[] = [
+      ofType(conditionallyUnlockItem.type),
+      withLatestFrom(state$),
+
+      filter(
```
```suggestion
      withLatestFrom(state$),
      filter(
```

rart commented on ui/app/src/state/epics/content.ts:
```diff
@@ -201,6 +202,31 @@ const content: CrafterCMSEpic[] = [
+      switchMap(([{ payload }, state]) => {
+        return unlock(state.sites.active, payload.path).pipe(
+          map(() =>
+            payload.notify === false
```
Would this cause two snacks?
Perhaps it'd better to only notify if notify === true

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+      {!readonly && (
+        <DialogFooter>
+          <Button onClick={onAddSnippet} endIcon={<ExpandMoreRoundedIcon />} className={classes.addSnippet}>
+            <FormattedMessage id="codeEditor.addCode" defaultMessage="Insert Code" />
```
```suggestion
            <FormattedMessage id="codeEditor.insertCode" defaultMessage="Insert Code" />
```

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+      <DialogBody className={classes.dialogBody}>
+        <ConditionalLoadingState isLoading={loading} classes={{ root: classes.loadingState }}>
+          <AceEditor
+            theme={`ace/theme/${type === 'light' ? 'chrome' : 'tomorrow_night'}`}
```
Let's move the theme switching logic to the AceEditor component. If theme prop is received, it uses that, if undefined, it does the theme.palette.type check to determine which of these to use.

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+            readOnly={disableEdit || readonly}
+            enableBasicAutocompletion={true}
+            enableSnippets={true}
+            enableLiveAutocompletion={true}
```
```suggestion
            enableBasicAutocompletion
            enableSnippets
            enableLiveAutocompletion
```

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+      });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [readonly, editorRef.current, content, disableEdit]);
```
Explain/comment why you're suppressing the exhaustive deps

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+              label: fields[key].name,
+              value: contentVariable.value.replace(
+                'VARIABLENAME',
+                hasUppercaseChars(fields[key].id) ? `["${dasherize(underscore(fields[key].id))}"]` : fields[key].id
```
Please explain this logic. What is it that the presence of uppercase chars dictates? What's happening here.

rart replied:
Logic not explained as requested. 👆

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+    }
+  }, [contentTypes, contentType, mode, item]);
+
+  const disableEdit = !item || (item.stateMap.locked && item.lockOwner !== user.username);
```
Would the util that we created for edit mode work here? We're not checking for write permission here, which perhaps we should be.

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,264 @@
+  onMinimized(): void;
+}
+
+export function CodeEditorDialogContainer(props: CodeEditorDialogContainerProps) {
```
Where is the item getting locked throughout opening it for edit? 
Do we need a conditionallyLockItem too?

aachavarria committed:
adding missing id for FormattedMessage
adding status: 500
removing space

aachavarria committed:
adding missing id for FormattedMessage
adding status: 500
removing space

aachavarria committed:
moving theme logic to aceEditor;

aachavarria committed:
removing editorRef.current from useEffect dependencies
updating conditionallyUnlockItem

aachavarria committed:
Creating localItemLock;

rart commented on ui/app/src/components/AceEditor.tsx:
```diff
@@ -181,10 +183,20 @@ export default React.forwardRef(function AceEditor(props: AceEditorProps, ref) {
+    { ...props, theme: props.theme ?? `ace/theme/${type === 'light' ? 'chrome' : 'tomorrow_night'}` } as AceOptions,
+    true,
+    ...aceOptions
+  );
```
```suggestion
  const options = pluckProps(props as AceOptions, true, ...aceOptions);
  options.theme = options.theme ?? `ace/theme/${type === 'light' ? 'chrome' : 'tomorrow_night'}`;
```

rart committed:
Update content.ts

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,262 @@
+            const snippets = Object.keys(fields).map((key) => ({
+              label: fields[key].name,
+              value: contentVariable.value.replace(
+                'VARIABLENAME',
```
```suggestion
                'VARIABLE_NAME',
```

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,262 @@
+              label: fields[key].name,
+              value: `${contentVariable.value}.${
+                // On groovy if  the ID has uppercase chars should be dasherize, example: fileName -> file-name
+                hasUppercaseChars(fields[key].id) ? `"${dasherize(underscore(fields[key].id))}"` : fields[key].id
```
Don't understand. Call me to review.

rart commented on ui/app/src/components/CodeEditorDialog/CodeEditorDialogContainer.tsx:
```diff
@@ -0,0 +1,262 @@
+        onDismiss={onClose}
+        rightActions={[
+          {
+            icon: 'MinimizeIcon',
```
Icon is a string?

aachavarria replied:
can be a string or icon

rart replied:
But if a string, doesn't it have to be an id? Like @material-ui/icons/Minimize? Otherwise, where/how are the string options defined?

aachavarria committed:
adding itemsBeingFetchedByPath

aachavarria committed:
Merge remote-tracking branch 'aaron/feature/4796' into feature/4796

aachavarria committed:
reading freemarkerCodeSnippets and groovyCodeSnippets from ui.config
adding datasets to uiconfig;

aachavarria committed:
renaming VARIABLENAME to VARIABLE_NAME

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4796

 Conflicts:
	ui/app/src/components/AceEditor.tsx
	ui/app/src/components/SystemStatus/GlobalDialogManager.tsx
	ui/app/src/utils/hooks.ts

aachavarria committed:
Fixing merge conflicts

aachavarria committed:
creating contentTypePropsMap

rart commented on ui/app/src/services/configuration.ts:
```diff
@@ -288,6 +289,17 @@ export function fetchSiteUiConfig(site: string): Observable<Pick<GlobalState['ui
         }
+        // reading datasets
+        xml.querySelectorAll('dataset').forEach((tag) => {
+          config.datasets[tag.id.replace('craftercms.', '')] = applyDeserializedXMLTransforms(
```
```suggestion
          config.datasets[tag.id] = applyDeserializedXMLTransforms(
```

rart commented on ui/app/src/components/AceEditor.tsx:
```diff
@@ -241,12 +254,24 @@ export default React.forwardRef(function AceEditor(props: AceEditorProps, ref) {
+      style={{
+        display: 'contents'
       }}
-      className={props.className ?? classes.base}
```
add prop `classes.{root,editorRoot}`.
please create class key and Styles typing

rart commented on ui/app/src/components/AceEditor.tsx:
```diff
@@ -241,12 +254,24 @@ export default React.forwardRef(function AceEditor(props: AceEditorProps, ref) {
+      }}
+      style={{
+        display: 'contents'
       }}
```
Move style on to `root` class

aachavarria committed:
adding root and editorRoot to AceEditorClassKey
creating classes on AceEditorProps

aachavarria committed:
Fixing issues relating with AceEditor classes changes
